python/chronos/src/bigdl/chronos/data/utils/prometheus_df.py

```diff
@@ -0,0 +1,193 @@
+                pd.Series((np.float64(v[1]) for v in r['values']),
+                          index=(pd.Timestamp(v[0], unit='s') for v in r['values']))
+            for r in data['result']})
+    elif result_type == 'scalar':
+        return np.float64(data['result'])
+    elif result_type == 'string':
+        return data['result']
+    else:
+        invalidInputError(False, f"The collected Prometheus data is unknown type {result_type}.")
+
+
+def metric_name(metric: dict) -> str:
+    """Convert metric labels to standard form."""
+    name = metric.get('__name__', '')
+    labels = ','.join(('{}={}'.format(k, json.dumps(v)) for k, v in metric.items()
+                      if k != '__name__'))
+    return '{0}{{{1}}}'.format(name, labels)
+
+
+def _timestamp(value):
+    if isinstance(value, datetime.datetime):
+        return value.timestamp()
+    else:
+        return value
+
+
+def _duration(value):
+    if isinstance(value, datetime.timedelta):
+        return value.total_seconds()
+    else:
+        return value
+
+
+def GetRangeDataframe(prometheus_url, query, starttime, endtime, step, columns, **kwargs):
+    '''
+    Convert the Prometheus data over the specified time period to dataframe and confirm
+    dt_col, target_col, id_col and extra_feature_col.
+
+    Return dataframe and col names.
+    '''
+    from bigdl.chronos.data.utils.utils import _check_type
+    _check_type(prometheus_url, "prometheus_url", str)
+    _check_type(query, "query", str)
+    _check_type(starttime, "starttime", (str, float))
+    _check_type(endtime, "endtime", (str, float))
+    _check_type(step, "step", (str, float))
+
+    # Generate dateframe
+    Pro_client = Prometheus(prometheus_url)
+    Pro_df = Pro_client.query_range(query, starttime, endtime, step, **kwargs)
+    df = pd.DataFrame(columns=Pro_df.columns.tolist())
+    df.insert(0, "datetime", Pro_df.index)
+    for col in Pro_df.columns.tolist():
+        df[col] = Pro_df[col].values
+
+    # Clean columns
+    output_columns = {"dt_col": "datetime",
+                      "target_col": Pro_df.columns.tolist(),
+                      "id_col": None,
+                      "extra_feature_col": None}
+    output_col_list = ["datetime"]
+    for col in ["target_col", "id_col", "extra_feature_col"]:
+        invalidInputError(columns[col] == [] or set(columns[col]).issubset(df.columns.tolist()),
+                          "The input " + col + " is not found in collected Prometheus data.")
+        if columns[col] != []:
+            if len(columns[col]) == 1:
+                output_columns[col] = columns[col][0]  # id_col is str
+            else:
+                output_columns[col] = columns[col]
+        if output_columns[col] is not None:
+            if isinstance(output_columns[col], list):
+                output_col_list = output_col_list + output_columns[col]
+            else:
+                output_col_list.append(output_columns[col])
+
+    df = df.loc[:, output_col_list]
```

Comment:
does `output_col_list` contains user input `extra_feature_col`?

Reply:
yes, the `output_col_list` is confirmed according to `output_columns`. `output_columns[extra_feature_col]` defaults to be None and will be updated if user specified `extra_feature_col`.

Reply:
okay I see, yeah it should be correct